Yosemite/Yosemite/Stores/ProductCategoryStore.swift

```diff
@@ -32,19 +36,64 @@ public final class ProductCategoryStore: Store {
         let remote = ProductCategoriesRemote(network: network)
 
-        remote.loadAllProductCategories(for: siteID, pageNumber: pageNumber, pageSize: pageSize) { (productCategories, error) in
+        remote.loadAllProductCategories(for: siteID, pageNumber: pageNumber, pageSize: pageSize) { [weak self] (productCategories, error) in
             guard let productCategories = productCategories else {
-                onCompletion(nil, error)
+                onCompletion(error)
                 return
             }
 
-            // TODO-2000 Insert product categories into Storage.framework
-            onCompletion(productCategories, nil)
+            self?.upsertStoredProductCategoriesInBackground(productCategories, siteID: siteID) {
+                onCompletion(nil)
+            }
+        }
+    }
+}
+
+// MARK: - Storage: ProductCategory
+//
+private extension ProductCategoryStore {
+    /// Updates (OR Inserts) the specified ReadOnly ProductCategory Entities *in a background thread*.
+    /// onCompletion will be called on the main thread!
+    ///
+    func upsertStoredProductCategoriesInBackground(_ readOnlyProductCategories: [Networking.ProductCategory],
+                                                   siteID: Int64,
+                                                   onCompletion: @escaping () -> Void) {
+        let derivedStorage = sharedDerivedStorage
+        derivedStorage.perform { [weak self] in
+            self?.upsertStoredProductCategories(readOnlyProductCategories, in: derivedStorage, siteID: siteID)
+        }
+
+        storageManager.saveDerivedType(derivedStorage: derivedStorage) {
+            DispatchQueue.main.async(execute: onCompletion)
+        }
+    }
+}
+
+private extension ProductCategoryStore {
+    /// Updates (OR Inserts) the specified ReadOnly ProductCategory entities into the Storage Layer.
+    ///
+    /// - Parameters:
+    ///     - readOnlyProducCategories: Remote ProductCategories to be persisted.
+    ///     - storage: Where we should save all the things!
+    ///     - siteID: site ID for looking up the ProductCategory.
+    ///
+    func upsertStoredProductCategories(_ readOnlyProductCategories: [Networking.ProductCategory],
+                                       in storage: StorageType,
+                                       siteID: Int64) {
+        // Upserts the ProductCategory models from the read-only version
+        for readOnlyProductCategory in readOnlyProductCategories {
+            let storageProductCategory: Storage.ProductCategory = {
+                if let storedCategory = storage.loadProductCategory(siteID: siteID, categoryID: readOnlyProductCategory.categoryID) {
+                    return storedCategory
+                }
+                return storage.insertNewObject(ofType: Storage.ProductCategory.self)
+            }()
+            storageProductCategory.update(with: readOnlyProductCategory)
```

Comment:
> I feel like a method for the individual ProductCategory would be more useful.

Hmmm 🤔, I guess it may be more useful or would made more sense if the class semantic was all around handling a individual category, however everything seems to be around a group of categories.  However I have no problem on go either way!

----

> Anyway. How did you feel when writing these? Any suggestions? 
> I'd really like these DB update methods outside of the *Store classes but I'm not sure yet what architectural pattern to introduce for that

Do you mean, to not do a `insert` directly in these classes? 
It does feel a bit risky that anyone can insert/delete/update to the underlying store and mess with the data integrity.

What I do normally is to have a `DataStore` type that sits in the outer layer of the `Storage.Framework` and exposes a single way that consumers(`Yosemite` in this case) can interact with a given data type. 

Let's suppose that a product can never be without a category. A `DataStore` would handle that integrity
```swift
final class ProductCatgoriesDataStore {
    func remove(category: ProductCategory) {
        let products = category.products
        storage.remove(category)

        for product in products {
            if product.categories.isEmpty {
                 let defaultCategory = storage.findCategory(id: "uncategorized")
                 product.categories.add(defaultCategory)
                 storage.update(product)
            }
        }
    }
}
```

That way, we stop providing a direct link to our underlying store.